src/auth/src/credentials/external_account.rs

```diff
@@ -0,0 +1,401 @@
+    ///
+    /// [quota project]: https://cloud.google.com/docs/quotas/quota-project
+    pub fn with_quota_project_id<S: Into<String>>(mut self, quota_project_id: S) -> Self {
+        self.quota_project_id = Some(quota_project_id.into());
+        self
+    }
+
+    /// Overrides the [scopes] for this credentials.
+    ///
+    /// [scopes]: https://developers.google.com/identity/protocols/oauth2/scopes
+    pub fn with_scopes<I, S>(mut self, scopes: I) -> Self
+    where
+        I: IntoIterator<Item = S>,
+        S: Into<String>,
+    {
+        self.scopes = Some(scopes.into_iter().map(|s| s.into()).collect());
+        self
+    }
+
+    /// Returns a [Credentials] instance with the configured settings.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [CredentialsError] if the `external_account_config`
+    /// provided to [`Builder::new`] cannot be successfully deserialized into the
+    /// expected format for an external account configuration. This typically happens if the
+    /// JSON value is malformed or missing required fields. For more information,
+    /// on the expected format, consult the relevant section in the
+    /// [external_account_credentials] guide.
+    ///
+    /// [external_account_credentials]: https://google.aip.dev/auth/4117#configuration-file-generation-and-usage
+    pub fn build(self) -> BuildResult<Credentials> {
+        let external_account_config: ExternalAccountConfig =
+            serde_json::from_value(self.external_account_config).map_err(BuilderError::parsing)?;
+
+        let mut config = external_account_config.clone();
+        if let Some(scopes) = self.scopes {
+            config.scopes = Some(scopes);
+        }
+
+        let token_provider = ExternalAccountTokenProvider {
+            subject_token_provider: external_account_config.credential_source,
+            config,
+        };
+
+        Ok(Credentials {
+            inner: Arc::new(ExternalAccountCredentials {
+                quota_project_id: self.quota_project_id.clone(),
+                token_provider: TokenCache::new(token_provider),
+            }),
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl<T> CredentialsProvider for ExternalAccountCredentials<T>
+where
+    T: CachedTokenProvider,
+{
+    async fn headers(&self, extensions: Extensions) -> Result<CacheableResource<HeaderMap>> {
+        let token = self.token_provider.token(extensions).await?;
+        build_cacheable_headers(&token, &self.quota_project_id)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use serde_json::*;
+
+    #[tokio::test]
+    async fn create_external_account_builder() {
+        let contents = json!({
+            "type": "external_account",
+            "audience": "//iam.googleapis.com/projects/<PROJECT_ID>/locations/global/workloadIdentityPools/<WORKLOAD_IDENTITY_POOL>/providers/<WORKLOAD_IDENTITY_PROVIDER_ID>",
+            "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
```

Comment:
Replace these long placeholders with something less distracting:

```suggestion
            "audience": "test-only-placeholder-audience",
            "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
```

Reply:
done